PostgreSQL 14 support

Hi,

pg_auto_failover 1.5.2 fails to build against v14. Can you please take a look? Thanks!

```
> gcc -std=c99 -D_GNU_SOURCE -g -I /usr/pgsql-14/include -I /usr/pgsql-14/include/server -I /usr/pgsql-14/include/internal -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wformat -Wall -Werror=implicit-int -Werror=implicit-function-declaration -Werror=return-type -Wno-declaration-after-statement -Wno-missing-braces -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/pg -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/log/src/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/subcommands.c/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/libs/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/parson/  -c -MMD -MP -MF/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl//.deps/cli_do_show.Po -o cli_do_show.o cli_do_show.c
In file included from group_state_machine.c:24:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
cat pgautofailover.sql > pgautofailover--1.5.sql
In file included from replication_state.c:19:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
gcc -std=c99 -D_GNU_SOURCE -g -I /usr/pgsql-14/include -I /usr/pgsql-14/include/server -I /usr/pgsql-14/include/internal -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wformat -Wall -Werror=implicit-int -Werror=implicit-function-declaration -Werror=return-type -Wno-declaration-after-statement -Wno-missing-braces -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/pg -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/log/src/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/subcommands.c/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/libs/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/parson/  -c -MMD -MP -MF/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl//.deps/cli_do_tmux.Po -o cli_do_tmux.o cli_do_tmux.c
In file included from metadata.c:18:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
/usr/lib64/ccache/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -O2  -std=c99 -Wall -Werror -Wno-unused-parameter -Iinclude -I/usr/pgsql-14/include -g -I. -I./ -I/usr/pgsql-14/include/server -I/usr/pgsql-14/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include -flto=thin -emit-llvm -c -o conninfo.bc conninfo.c
make[2]: *** [<builtin>: group_state_machine.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from pg_auto_failover.c:19:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
gcc -std=c99 -D_GNU_SOURCE -g -I /usr/pgsql-14/include -I /usr/pgsql-14/include/server -I /usr/pgsql-14/include/internal -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wformat -Wall -Werror=implicit-int -Werror=implicit-function-declaration -Werror=return-type -Wno-declaration-after-statement -Wno-missing-braces -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/pg -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/log/src/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/subcommands.c/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/libs/ -I/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl/../lib/parson/  -c -MMD -MP -MF/home/devrim/Documents/Devrim/Projects/repo/pgrpms/rpm/redhat/master/non-common/pg_auto_failover/master/pg_auto_failover-1.5.2/src/bin/pg_autoctl//.deps/cli_do_tmux_azure.Po -o cli_do_tmux_azure.o cli_do_tmux_azure.c
In file included from health_check_worker.c:18:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
In file included from node_metadata.c:20:
version_compat.h:19:2: error: #error "Unknown or unsupported postgresql version"
   19 | #error "Unknown or unsupported postgresql version"
      |  ^~~~~
make[2]: *** [<builtin>: replication_state.o] Error 1
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -18,18 +18,22 @@ matrix:
     - env: PGVERSION=11 TEST=multi
     - env: PGVERSION=12 TEST=multi
     - env: PGVERSION=13 TEST=multi
+    - env: PGVERSION=14 TEST=multi
     - env: PGVERSION=10 TEST=single
     - env: PGVERSION=11 TEST=single
     - env: PGVERSION=12 TEST=single
     - env: PGVERSION=13 TEST=single
+    - env: PGVERSION=14 TEST=single
     - env: PGVERSION=10 TEST=ssl
     - env: PGVERSION=11 TEST=ssl
     - env: PGVERSION=12 TEST=ssl
     - env: PGVERSION=13 TEST=ssl
+    - env: PGVERSION=14 TEST=ssl
     - env: PGVERSION=10 TEST=monitor
     - env: PGVERSION=11 TEST=monitor
     - env: PGVERSION=12 TEST=monitor
     - env: PGVERSION=13 TEST=monitor
+    - env: PGVERSION=14 TEST=monitor
     - env: LINTING=true
 before_install:
   - git clone -b v0.7.18 --depth 1 https://github.com/citusdata/tools.git
@@ -40,6 +44,8 @@ before_install:
   - pyenv versions
   - pyenv install -s 3.7
   - pyenv global 3.7
+  - sudo apt-get install liblz4-1
+  - sudo apt-get install liblz4-dev
   - sudo apt-get install bridge-utils
   - sudo apt-get install python3-pip
   - sudo pip3 install --upgrade pip

--- a/.travis.yml
+++ b/.travis.yml
@@ -24,16 +24,16 @@ matrix:
     - env: PGVERSION=12 TEST=single
     - env: PGVERSION=13 TEST=single
     - env: PGVERSION=14 TEST=single
-    - env: PGVERSION=10 TEST=ssl
-    - env: PGVERSION=11 TEST=ssl
-    - env: PGVERSION=12 TEST=ssl
-    - env: PGVERSION=13 TEST=ssl
-    - env: PGVERSION=14 TEST=ssl
     - env: PGVERSION=10 TEST=monitor
     - env: PGVERSION=11 TEST=monitor
     - env: PGVERSION=12 TEST=monitor
     - env: PGVERSION=13 TEST=monitor
     - env: PGVERSION=14 TEST=monitor
+    - env: PGVERSION=10 TEST=ssl
+    - env: PGVERSION=11 TEST=ssl
+    - env: PGVERSION=12 TEST=ssl
+    - env: PGVERSION=13 TEST=ssl
+    - env: PGVERSION=14 TEST=ssl
     - env: LINTING=true
 before_install:
   - git clone -b v0.7.18 --depth 1 https://github.com/citusdata/tools.git

--- a/Makefile
+++ b/Makefile
@@ -15,13 +15,15 @@ TESTS_MONITOR += test_installcheck
 TESTS_MONITOR += test_monitor_disabled
 TESTS_MONITOR += test_replace_monitor
 
+# This could be in TESTS_MULTI, but adding it here optimizes Travis run time
+TESTS_MONITOR += test_multi_alternate_primary_failures
+
 # Tests for single standby
 TESTS_SINGLE  = test_auth
 TESTS_SINGLE += test_basic_operation
 TESTS_SINGLE += test_basic_operation_listen_flag
 TESTS_SINGLE += test_create_run
 TESTS_SINGLE += test_create_standby_with_pgdata
-TESTS_SINGLE += test_debian_clusters
 TESTS_SINGLE += test_ensure
 TESTS_SINGLE += test_skip_pg_hba
 
@@ -30,12 +32,14 @@ TESTS_SSL  = test_enable_ssl
 TESTS_SSL += test_ssl_cert
 TESTS_SSL += test_ssl_self_signed
 
+# This could be in TESTS_SINGLE, but adding it here optimizes Travis run time
+TESTS_SSL += test_debian_clusters
+
 # Tests for multiple standbys
 TESTS_MULTI  = test_multi_async
 TESTS_MULTI += test_multi_ifdown
 TESTS_MULTI += test_multi_maintenance
 TESTS_MULTI += test_multi_standbys
-TESTS_MULTI += test_multi_alternate_primary_failures
 
 # TEST indicates the testfile to run
 TEST ?=

--- a/src/bin/pg_autoctl/Makefile
+++ b/src/bin/pg_autoctl/Makefile
@@ -54,7 +54,7 @@ endif
 override CFLAGS := $(DEFAULT_CFLAGS) $(CFLAGS)
 
 LIBS  = -L $(shell $(PG_CONFIG) --pkglibdir)
-LIBS  = -L $(shell $(PG_CONFIG) --libdir)
+LIBS += -L $(shell $(PG_CONFIG) --libdir)
 LIBS += $(shell $(PG_CONFIG) --ldflags)
 LIBS += $(shell $(PG_CONFIG) --libs)
 LIBS += -lpq

--- a/src/bin/pg_autoctl/Makefile
+++ b/src/bin/pg_autoctl/Makefile
@@ -54,6 +54,7 @@ endif
 override CFLAGS := $(DEFAULT_CFLAGS) $(CFLAGS)
 
 LIBS  = -L $(shell $(PG_CONFIG) --pkglibdir)
+LIBS  = -L $(shell $(PG_CONFIG) --libdir)
 LIBS += $(shell $(PG_CONFIG) --ldflags)
 LIBS += $(shell $(PG_CONFIG) --libs)
 LIBS += -lpq

--- a/src/bin/pg_autoctl/cli_common.c
+++ b/src/bin/pg_autoctl/cli_common.c
@@ -1337,7 +1337,7 @@ keeper_cli_print_version(int argc, char **argv)
 				"pg_autoctl extension version %s\n",
 				PG_AUTOCTL_EXTENSION_VERSION);
 		fformat(stdout, "compiled with %s\n", PG_VERSION_STR);
-		fformat(stdout, "compatible with Postgres 10, 11, 12, and 13\n");
+		fformat(stdout, "compatible with Postgres 10, 11, 12, 13, and 14\n");
 	}
 
 	exit(0);

--- a/src/bin/pg_autoctl/cli_do_tmux.c
+++ b/src/bin/pg_autoctl/cli_do_tmux.c
@@ -490,7 +490,20 @@ tmux_setenv(PQExpBuffer script,
 			const char *sessionName, const char *root, int firstPort)
 {
 	char PATH[BUFSIZE] = { 0 };
+	char PG_CONFIG[MAXPGPATH] = { 0 };
 	char monitor_pguri[MAXCONNINFO] = { 0 };
+
+	if (env_exists("PG_CONFIG") &&
+		!get_env_copy("PG_CONFIG", PG_CONFIG, sizeof(PG_CONFIG)))
+	{
+		log_fatal("Failed to get PG_CONFIG from the environment");
+		exit(EXIT_CODE_INTERNAL_ERROR);
+	}
+
+	tmux_add_command(script,
+					 "set-environment -t %s PG_CONFIG \"%s\"",
+					 sessionName,
+					 PG_CONFIG);
 
 	if (!get_env_copy("PATH", PATH, sizeof(PATH)))
 	{

--- a/src/bin/pg_autoctl/cli_do_tmux.c
+++ b/src/bin/pg_autoctl/cli_do_tmux.c
@@ -493,17 +493,19 @@ tmux_setenv(PQExpBuffer script,
 	char PG_CONFIG[MAXPGPATH] = { 0 };
 	char monitor_pguri[MAXCONNINFO] = { 0 };
 
-	if (env_exists("PG_CONFIG") &&
-		!get_env_copy("PG_CONFIG", PG_CONFIG, sizeof(PG_CONFIG)))
+	if (env_exists("PG_CONFIG"))
 	{
-		log_fatal("Failed to get PG_CONFIG from the environment");
-		exit(EXIT_CODE_INTERNAL_ERROR);
-	}
+		if (!get_env_copy("PG_CONFIG", PG_CONFIG, sizeof(PG_CONFIG)))
+		{
+			log_fatal("Failed to get PG_CONFIG from the environment");
+			exit(EXIT_CODE_INTERNAL_ERROR);
+		}
 
-	tmux_add_command(script,
-					 "set-environment -t %s PG_CONFIG \"%s\"",
-					 sessionName,
-					 PG_CONFIG);
+		tmux_add_command(script,
+						 "set-environment -t %s PG_CONFIG \"%s\"",
+						 sessionName,
+						 PG_CONFIG);
+	}
 
 	if (!get_env_copy("PATH", PATH, sizeof(PATH)))
 	{

--- a/src/bin/pg_autoctl/parsing.c
+++ b/src/bin/pg_autoctl/parsing.c
@@ -195,6 +195,12 @@ parse_dotted_version_string(const char *pg_version_string, int *pg_version)
 		}
 	}
 
+	/* Postgres alpha/beta versions report version "14" instead of "14.0" */
+	if (!dotFound)
+	{
+		strlcpy(minor, "0", INTSTRING_MAX_DIGITS);
+	}
+
 	int maj = 0;
 	int min = 0;
 

--- a/src/monitor/version_compat.h
+++ b/src/monitor/version_compat.h
@@ -14,8 +14,8 @@
 
 #include "postgres.h"
 
-/* we only support Postgres versions 10, 11, 12 and 13. */
-#if (PG_VERSION_NUM < 100000 || PG_VERSION_NUM >= 140000)
+/* we support Postgres versions 10, 11, 12, 13, and 14. */
+#if (PG_VERSION_NUM < 100000 || PG_VERSION_NUM >= 150000)
 #error "Unknown or unsupported postgresql version"
 #endif
 

